### PR TITLE
Add linters: isort, pycodestyle, pydocstyle, pylint

### PR DESCRIPTION
**Pull request by virtualtam**
### isort
isort can be used:
- as a linter, to check import statements follow a given coding style
- to automatically reformat import statements according to a given set of rules

The recommended ordering is:
- standard library imports
- related third party imports
- local application/library specific imports

Example:
```python
import os
from datetime import datetime

import requests

from mypackage.utils import some_function         # explicit absolute import (recommended)
from .utils import some_other_function, ToolClass # explicit relative import

[...]
```

See:
- https://github.com/timothycrosley/isort
- https://www.python.org/dev/peps/pep-0008/#imports

### pycodestyle
`pycodestyle` (formerly `pep8`) enforces the PEP-8 style guide to ensure code
follows the base Python conventions.

See:
- https://pycodestyle.readthedocs.io/en/latest/
- https://www.python.org/dev/peps/pep-0008/

### pydocstyle
`pydocstyle` checks module, class, method and function docstring against
the PEP-257 style guide and helps keeping a consistent documentation style
across sources.

Default disabled checks:
- D203 - 1 blank line required before class docstring
- D212 - Multi-line docstring summary should start at the first line
- D213 - Multi-line docstring summary should start at the second line
- D404 - First word of the docstring should not be This

Disabled checks:
- D105 - Missing docstring in magic method
- D400 - First line should end with a period
- D401 - First line should be in imperative mood

See:
- http://www.pydocstyle.org/en/latest/
- http://www.pydocstyle.org/en/latest/usage.html
- http://www.pydocstyle.org/en/latest/error_codes.html
- http://www.python.org/dev/peps/pep-0257/
- https://github.com/python/peps/blob/master/pep-0257.txt

### pylint
`pylint` checks Python code using numerous heuristics, and is of great help
when dealing with a dynamically typed language.

Some of Pylint's features:
- PEP-8 (as a complement to pycodestyle)
- class, function, method, variable naming
- code complexity:
  - branches
  - statements
- module depreciation
- code duplication
- graphviz/dot UML diagram generation

All checks can be configured through the `pylintrc` configuration file,
and it is possible to locally disable specific checks when needed:

```python
def my_function_with_an_exceptionally_long_name(arg1, arg2):
    """Some useless function"""
    # pylint: disable=unused-argument,invalid-name
    pass
```

It is also possible to use several configuration files to fit different usages:
- human-readable output: colorized text or HTML
- machine-parseable-output for CI services

Notes:
- Python 3.6 is not (yet) supported by a released pylint version
- hence, pylint execution is controlled by a Tox condition

See:
- https://www.pylint.org/
- https://pylint.readthedocs.io/en/latest/
- https://pylint.readthedocs.io/en/latest/user_guide/options.html
- https://pylint.readthedocs.io/en/latest/user_guide/ide-integration.html
- https://www.logilab.org/blogentry/6883
- https://github.com/carsongee/pytest-pylint
- https://tox.readthedocs.io/en/latest/config.html

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -1,5 +1,14 @@
...
+envlist=py3{4,5,6},docs
+skip_missing_interpreters=True
+
+[testenv]
+deps=-rrequirements/tests.txt
+commands=
+    isort --check-only --diff -p shaarli_client
+    pycodestyle
+    pydocstyle
+    py34,py35: pytest --pylint
```
Comment (nodiscc) [inline tox.ini]: For reference:


    Python 3.6 is not (yet) supported by a released pylint version
    hence, pylint execution is controlled by a Tox condition

File: setup.cfg [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+skip=.git,.tox,docs
+
+[pycodestyle]
+count=True
+exclude=.git,.tox,docs,__pycache__
+max-line-length=80
+statistics=True
+
+[pydocstyle]
+ignore=D105,D400,D401
```
Comment (nodiscc) [inline setup.cfg]: For reference: 
http://www.pydocstyle.org/en/latest/error_codes.html

    Missing docstrings/Docstring Content Issues
    D105 	Missing docstring in magic method
    D400 	First line should end with a period
    D401 	First line should be in imperative mood

**Comment by nodiscc** [comment]
Thank you :+1:

**Comment by virtualtam** [comment]
Thanks @nodiscc for reviewing these PRs!
I'll take time to add tool-related information to the Sphinx docs so we have it handy for devs ;-)

**Comment by ArthurHoaro** [comment]
If I understand, these tools are here used to check code style (and other stuff) in CI, so to check PRs. Is there a possibility to apply the same kind of thing to Shaarli, only regarding the diff?

**Comment by virtualtam** [comment]
@ArthurHoaro that'd be helpful; we have the Makefile targets introduced starting with https://github.com/shaarli/Shaarli/issues/95 but the output is quite verbose, even when scans are restricted to `application`, `tests` and `plugins` dirs.

We could start with fixing warnings issued by PHP Mess Detector (~500) for these dirs and then add the corresponding Makefile target to Travis.

Another option is to use a CI linting service such as [SensioLabs Insight](https://insight.sensiolabs.com/) or [Code Climate](https://codeclimate.com/)

I've enabled the former on my fork, see results at https://insight.sensiolabs.com/projects/3f368e16-d9df-409c-b186-854cab3036e6 ; I think I have an account on CodeClimate but never took the time to delve into configuring scans for Shaarli :)

**Comment by ArthurHoaro** [comment]
Well, yes, we can't check the whole codebase. Even the folders the community have created, it should have been done since the beginning. And we can't fix all those style issues while we're trying to bring huge changes to Shaarli architecture and code base.

My question was more: along all these awesome tools, isn't there anyone which support (or any alternative way to support) only pull request's diff? This way, we'll be sure to be improving existing, sometime messy, code without breaking the existent one. 

I'm asking for your experience, but I'll take a look later ; I'm pretty sure a lot of companies have had the same issue.

**Comment by virtualtam** [comment]
I'm not sure it's widely spread, though I've been asked such a feature (diff-only lint) quite often to be enabled on CI platforms ;-)

The use case is mainly when you want to check your own contributions over a (huge) forked codebase and you don't want to focus on previous code.

The Linux kernel has such a tool to check coding style for incoming contribs, and I guess it could be do-able by feeding commit diffs to existing tools, yet with several shortcomings:
- no/lack of context to spot missing imports and unused variables
- no scoping e.g. if you don't know the lines being analyzed come from a function/method/globals

Experience: watching a Linux vendor gatekeepers become mad after being flooded by Gerrit patches from rookie devs randomly attempting to fix each and every detected error ^^

For the sake of it: https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/scripts/checkpatch.pl

What I'd suggest is to enable a tool to have regular code sanity reports, to avoid introducing new bugs and help fixing the easy ones. I've found SL Insight a bit limited in terms of reporting and integration, Code Climate on the other hand seems quite feature-complete and is worth investigating.

**Comment by ArthurHoaro** [comment]
Yes, you're right, that was a false good idea. Checking out of context code, even only for coding style, doesn't really make sense. Also, we already have a `make` target which can help (but I didn't run it for a long time tbh).

**Pull request merged**